### PR TITLE
Week 5 assessment elmer figueroa

### PR DESCRIPTION
**Pull request by drfudd2001**
# Submitting Week 5 Assessment

### ✅ Completed Files

Mark the following as confirmation that the files in this repository are complete.

- [x] JavaScript Coding Challenges
- [x] Ruby Coding Challenges
- [x] Interview Questions

### 📝 Feedback

If you would like specific feedback on a particular problem please leave details here:
Hope the video game references are welcomed

### CODE REVIEW

File: code-challenges.test.js [hunk 1]
```diff
@@ -21,8 +21,73 @@ const secretCodeWord2 = "Gobbledygook"
...
+const codedMessage = (string) => {
+    const codeReplacements = { 'a': '4', 'e': '3', 'i': '1', 'o': '0' }
+    let codedSolution = ''  
+    for (const letter of string) {
+        const lowerCaseLetter = letter.toLowerCase()
+        codedSolution += lowerCaseLetter in codeReplacements ? codeReplacements[lowerCaseLetter] : letter
+    }
+    return codedSolution
+}
+
```
Comment (chrisyhlee) [inline code-challenges.test.js:63]: Beautiful!

File: code-challenges.test.js [hunk 2]
```diff
@@ -21,8 +21,73 @@ const secretCodeWord2 = "Gobbledygook"
...
+//         const lowerCaseLetter = letter.toLowerCase()
+//         codedSolution += codeReplacements[lowerCaseLetter] || letter
+//     }
+//     return codedSolution
+// }
+
+// Uses the logical OR (||) operator to provide a default value of letter when codeReplacements[lowerCaseLetter] is falsy (i.e., undefined).
+// This eliminates the need to use the in operator to check if lowerCaseLetter exists as a key in codeReplacements object.
+// This results in a simpler and more concise code while achieving the same result as the original code.
+
```
Comment (chrisyhlee) [inline code-challenges.test.js:88]: Beautiful-er refactor! Bonus research: check out `.replace()`!

File: code-challenges.test.js [hunk 3]
```diff
@@ -34,19 +99,145 @@ const letterA = "a"
...
+    // Tests:       2 passed, 2 total
+    // Snapshots:   0 total
+    // Time:        0.155 s, estimated 1 s
+    // Ran all test suites.
+
+// Refactor:
+// const doesItContainLetter = (array, letter) => {
+//     return array.filter(word => word.toLowerCase().includes(letter.toLowerCase()))
+// }
+
```
Comment (chrisyhlee) [inline code-challenges.test.js:151]: Nice job catching the unnecessary variable and going back for the refactor! And way to handle edge cases!

File: code-challenges.test.js [hunk 4]
```diff
@@ -34,19 +99,145 @@ const letterA = "a"
...
+    for (const cardNum in winningHand) {
+        if (winningHand[cardNum] === 2) {
+            hasPair = true
+        } else if (winningHand[cardNum] === 3) {
+            hasThreeOfAKind = true
+        }
+    }
+    return hasPair && hasThreeOfAKind
+}
+
```
Comment (chrisyhlee) [inline code-challenges.test.js:229]: Well done! Love the use of an object as "memory"! Bonus challenge: check out `.reduce()`! By the way, I'm only adding these bonus research tips in case you wanted to look at them, by no means is it required!

File: interview-questions.md [hunk 5]
```diff
@@ -6,46 +6,169 @@ First, without external resources. Challenge yourself to answer from memory.
...
 
-1. What does calling super do in a Ruby class?
+---
+
+### 1. What does calling super do in a Ruby class?
 
 Your answer:
 
+In Ruby, super is like calling the same method in the parent class when you're in a subclass. It lets the subclass use the same method from the parent class and add extra behavior to it if needed. It's like borrowing something from your parent and adding your own touch to it.
+
```
Comment (chrisyhlee) [inline interview-questions.md:16]: Nice analogy

File: interview-questions.md [hunk 6]
```diff
@@ -6,46 +6,169 @@ First, without external resources. Challenge yourself to answer from memory.
...
+```
+In this example, the Player class has a jump method that prints "Player jumps". The Mario class inherits from the Player class and overrides the jump method. The super call within the Mario class invokes the jump method from the Player class, allowing Mario to extend the behavior of the parent class's method by adding additional code after the super call.
+
+When mario.jump is called, it will output:
+```ruby
+Mario jumps higher
+Player jumps
+```
+As you can see, the jump method from the Player class is called first due to the super call in the Mario class, followed by the additional code in the Mario class's jump method. This allows Mario to inherit the behavior of the parent class's method and extend it with its own modifications.
+
```
Comment (chrisyhlee) [inline interview-questions.md:45]: Nice example! Just a note: when `mario.jump` is called, the `jump` method of the `Mario` class is executed first, which is seen in the output when "Mario jumps higher" prints before "Player jumps"

File: interview-questions.md [hunk 7]
```diff
@@ -6,46 +6,169 @@ First, without external resources. Challenge yourself to answer from memory.
...
+For example, the 'rails' gem is a popular Ruby gem that provides a framework for building web applications. Just like how power-up items in Sonic the Hedgehog enhance Sonic's abilities, the 'rails' gem enhances a Ruby application's ability to build robust and scalable web applications by providing pre-built functions and conventions for developing web applications.
+
+Another example is the 'faker' gem, which is a popular Ruby gem for generating fake data for testing and development purposes. Similar to how power-up items in Sonic the Hedgehog give Sonic an advantage, the 'faker' gem gives developers an advantage by providing ready-made functions for generating realistic fake data in Ruby applications, saving time and effort in creating test data.
+
+In summary, just like power-up items in Sonic the Hedgehog enhance Sonic's abilities, gems in Ruby provide pre-built functionalities that can be added to Ruby projects to extend their capabilities and make development tasks easier and more efficient.
+
```
Comment (chrisyhlee) [inline interview-questions.md:63]: Love the Sonic analogies! Not a Sonic guy myself but I can respect it.

File: interview-questions.md [hunk 8]
```diff
@@ -6,46 +6,169 @@ First, without external resources. Challenge yourself to answer from memory.
...
+Network database: It's like a more complex tree-like structure in Cities: Skylines, where buildings and infrastructure can have multiple connections and relationships with other elements, similar to how different buildings may be interconnected with roads, power lines, or water pipes in the game.
+
+NoSQL database: It's like using different data models, such as key-value, document, column-family, or graph, to store and manage data in different ways in Cities: Skylines. For example, using a key-value model to store data on available resources or using a graph model to represent the relationships between different districts or neighborhoods in the city.
+
+In summary, a relational database is one way of organizing data, while other types of databases provide different approaches based on data models and structures, just like how different methods of organizing data are used in Cities: Skylines to manage the virtual city efficiently.
+
```
Comment (chrisyhlee) [inline interview-questions.md:87]: Great examples! You must love Cities: Skylines.....

File: interview-questions.md [hunk 9]
```diff
@@ -6,46 +6,169 @@ First, without external resources. Challenge yourself to answer from memory.
...
+Moreover, primary keys are crucial for efficient data retrieval and linking data across tables. Just like how Crash Bandicoot needs to quickly locate a specific crate using its unique identifier, databases use primary keys to quickly locate and retrieve specific rows of data. Primary keys also serve as reference points for establishing relationships between different tables in a relational database, allowing data to be linked and combined based on their unique identifiers.
+
+In summary, primary keys are like unique identifiers assigned to each row in a table in a relational database, similar to how crates have unique identifiers in Crash Bandicoot. They are important for maintaining data integrity, enabling efficient data retrieval, and establishing relationships between tables, just as they help identify and differentiate crates in the game.
+
```
Comment (chrisyhlee) [inline interview-questions.md:105]: Great answer! Can't go wrong with Crash

File: interview-questions.md [hunk 10]
```diff
@@ -6,46 +6,169 @@ First, without external resources. Challenge yourself to answer from memory.
...
+PUT: Corresponds to the "Update" action in CRUD. It updates an existing resource with new data. For example, when you modify a Sim's attributes or properties, such as changing their name or appearance, it would use the PUT method to update that information on the server.
+
+DELETE: Corresponds to the "Delete" action in CRUD. It requests to remove or delete a resource. For example, when you delete a Sim character from your game, it would use the DELETE method to send a request to the server to delete that character's data.
+
+PATCH: Corresponds to a partial update of a resource, where only the changes or modifications are sent to the server, instead of sending the entire resource. For example, when you update only the Sim's skill level or relationship status, it would use the PATCH method to send only the changes to the server, rather than sending the entire Sim's data.
+
+In summary, HTTP verbs are the standard set of actions used in the HTTP protocol to perform operations on resources, and their corresponding CRUD actions are: GET for Read, POST for Create, PUT for Update, DELETE for Delete, and PATCH for Partial Update.
+
```
Comment (chrisyhlee) [inline interview-questions.md:130]: Wow awesome explanation of HTTP verbs and their corresponding CRUD actions using Sims! This one's my favorite

File: interview-questions.md [hunk 11]
```diff
@@ -6,46 +6,169 @@ First, without external resources. Challenge yourself to answer from memory.
...
 ## Looking Ahead: Terms for Next Week
 
 Research and define the following terms to the best of your ability.
 
-1. Model validations:
+### 1. Model validations:
+
+Model validation is like a safety net for data in an application, such as in a video game like God of War or Final Fantasy. It ensures that the data entered or submitted by users or other parts of the application meets certain criteria or rules before it is accepted and stored in the system. This helps maintain data integrity, accuracy, and consistency. For example, it can prevent invalid data from being saved in the game, such as an invalid character name, an incorrect weapon type, or an invalid quest status. Model validation helps ensure that the data in the application is valid and reliable, reducing the risk of errors or inconsistencies that could impact the functionality or performance of the application.
+
```
Comment (chrisyhlee) [inline interview-questions.md:140]: Great explanation

File: ruby-challenges.rb [hunk 12]
```diff
@@ -1,32 +1,105 @@
...
 letter_o = 'o'
 # Expected output: ['coffee', 'soda water']
 letter_t = 't'
 # Expected output: ['tea', 'water', 'soda water']
 
+def words_containing_letter(words, letter)
+    filtered_words = words.select { |word| word.downcase.include?(letter.downcase) }
+    return filtered_words
+end
+
```
Comment (chrisyhlee) [inline ruby-challenges.rb:22]: Don't forget Ruby has implicit return! If you wanted to use the keyword `return`, you could just throw it in place of the variable!

File: ruby-challenges.rb [hunk 13]
```diff
@@ -1,32 +1,105 @@
...
+us_states = { northwest: ['Washington', 'Oregon', 'Idaho'], southwest: ['California', 'Arizona', 'Nevada'], northeast: ['Maine', 'New Hampshire', 'Rhode Island'] }
 # Expected output: ['Arizona', 'California', 'Idaho', 'Maine', 'Nevada', 'New Hampshire', 'Oregon', 'Rhode Island', 'Washington'] 
 
+def ordered_states_unleashed(hash)
+    values = hash.values.flatten
+    values.sort!
+    return values
+end
+
```
Comment (chrisyhlee) [inline ruby-challenges.rb:44]: Same as before, remember implicit return. Also, you can chain the `flatten` and `sort` methods

File: ruby-challenges.rb [hunk 14]
```diff
@@ -1,32 +1,105 @@
...
+    def pedal_faster(speed_increase)
+        @current_speed += speed_increase
+    end
+
+    def brake(speed_decrease)
+        @current_speed -= speed_decrease
+        @current_speed = [0, @current_speed].max
+    end
+end
+
```
Comment (chrisyhlee) [inline ruby-challenges.rb:91]: Great use of `.max`!

**Review by chrisyhlee — APPROVED** [review]
Fantastic job on assessment 5! This assessment displays Competent level! While there were a couple changes/methods you could make/use to make your code more efficient (which I left in the comments for you to look at), your understanding of the concepts behind them is definitely at the Competent level. Also, thanks for making this entertaining to review with the video game references! Keep up the great work!

File: interview-questions.md (hunk 8 above)
Comment (drfudd2001) [reply interview-questions.md:87]: I used to play it for hours at a time! 😅

File: ruby-challenges.rb (hunk 12 above)
Comment (drfudd2001) [reply ruby-challenges.rb:22]: Thank you for catching that. I was up late and when I got to the Ruby challenges, it was past 2am, I think 🙈... I will keep in mind.
